README
This PAM module switches the execution context of the calling process
to a running OpenVZ container, like the "vzctl enter" command.
It is tipically used by a daemon like ssh or cron running in the 
Host Node to transparently run itself and its child processes in a 
container.
It is not supposed to be used with "standard" OpenVZ installations:
the system must be designed to share some components between the HN
and the affected contexts.

To use this module you need to add a line like this to the PAM
configuration file of the relevant daemon:

session required pam_vz.so

Then the daemon will enter the context with the CTID set to the UID or
GID of the user.

Beware: the module cannot be used with su(1) and interactive sessions,
because the stdin/stdout/stderr file descriptors of the process will
still be the ones in the HN.
It also cannot work with FTP servers because they will try to open the
data connections using the context IP address instead of the HN one.

Valid options for the module are:

* debug
Enables verbose logging.

* min_id=NUMBER
* max_id=NUMBER
Define the inclusive range of UIDs/GIDs which will cause entering the
associated context.
Default: min_id=10000, max_id=65533

* use_as_ctid=[uid|gid]
Chooses to use either the UID or the GID as the CTID.
Default: use_as_ctid=uid

* ctid_delta=number
Adds or subtracts a number from the CTID.
Default: ctid_delta=0

* closelog
Closes the syslog socket after entering the CT.
This will break logging unless a syslog daemon is running in the CT.



debian/changelog
libpam-vz (0.01-1) unstable; urgency=low

  * Initial release.

 -- Marco d'Itri <md@linux.it>  Thu, 30 Jul 2009 04:34:47 +0200


ve_enter.c
/*
 * Copyright 2009 by Marco d'Itri <md@linux.it>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#define _GNU_SOURCE

/* System library */
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <string.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <errno.h>
#include <fcntl.h>

/* OpenVZ-related includes */
#include "linux/vzcalluser.h"
#include "vzsyscalls.h"
#include "types.h"

/* FIXME: this is the Debian default, other systems may use /vz/ */
#undef VZ_DIR
#define VZ_DIR "/var/lib/vz/"

/* Prototypes */
int ve_enter(const id_t veid);
static inline int setluid(const id_t uid);

#define log_e(x, ...) do { \
    syslog(LOG_AUTHPRIV|LOG_ERR, "pam_vz: " x, ## __VA_ARGS__); \
} while(0);

int ve_enter(const id_t veid)
{
    int vzfd;
    int errcode;
    int retry = 0;
    struct vzctl_env_create env_create;
    char *root;

    if ((vzfd = open(VZCTLDEV, O_RDWR)) < 0) {
	log_e("open(%s): %m", VZCTLDEV);
	return 0;
    }

    if (asprintf(&root, "%sroot/%u/", VZ_DIR, veid) == -1) {
	log_e("malloc: %m");
	close(vzfd);
	return 0;
    }

    if (chroot(root) < 0) {
	log_e("chroot(%s): %m", root);
	free(root);
	close(vzfd);
	return 0;
    }

    free(root);

    if (chdir("/") < 0) {
	log_e("chdir(/): %m");
	close(vzfd);
	return 0;
    }

    /* associate the process with its beancounter */
    if (setluid(veid) < 0) {
	log_e("setluid(%d): %m", veid);
	close(vzfd);
	return 0;
    }

    memset(&env_create, 0, sizeof(env_create));
    env_create.veid = veid;
    env_create.flags = VE_ENTER;

    /* enter the VE */
    do {
	if (retry)
	    sleep(1);
	errcode = ioctl(vzfd, VZCTL_ENV_CREATE, &env_create);
    } while (errcode < 0 && errno == EBUSY && retry++ < 3);

    close(vzfd);

    if (errcode < 0) {
	log_e("ioctl(VZCTL_ENV_CREATE): %m");
	return 0;
    }

    return 1;
}

/****************************************************************************/
static inline int setluid(const id_t uid)
{
    return syscall(__NR_setluid, uid);
}

/****************************************************************************/
#ifdef TEST

int main(int argc, char *argv[], char *env[])
{
    id_t veid = 10000;

    if (!*++argv) {
	fprintf(stderr, "Usage: ve_enter PROGRAM [ARG]...\n");
	exit(1);
    }

    if (!ve_enter(veid)) {
	fprintf(stderr, "ve_enter failed!\n");
	exit(1);
    }

    execv(argv[0], argv);

    fprintf(stderr, "exec(%s): %s\n", argv[0], strerror(errno));
    exit(1);
}

#endif



include/types.h
/*
 *  Copyright (C) 2000-2008, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#ifndef	_TYPES_H_
#define	_TYPES_H_

#define VZ_DIR			PKGCONFDIR "/"
#define GLOBAL_CFG		VZ_DIR "vz.conf"
#define VPS_CONF_DIR		VZ_DIR "conf/"
#define LIB_SCRIPTS_DIR		PKGLIBDIR "/scripts/"
#define DIST_DIR		VZ_DIR "dists"
#define VENAME_DIR		VZ_DIR "names"

#define VZCTLDEV		"/dev/vzctl"

#define VZFIFO_FILE		"/.vzfifo"

#define VPS_STOP		LIB_SCRIPTS_DIR "vps-stop"
#define VPS_NET_ADD		LIB_SCRIPTS_DIR "vps-net_add"
#define VPS_NET_DEL		LIB_SCRIPTS_DIR "vps-net_del"

#define envid_t 	unsigned int
#define STR_SIZE	512
#define PATH_LEN	256

#define NONE		0
#define YES		1
#define NO		2

#define ADD		0
#define DEL		1

/* Default environment variable PATH */
#define	ENV_PATH	"PATH=/bin:/sbin:/usr/bin:/usr/sbin:"

/* CT states */
enum {
	STATE_STARTING = 1,
	STATE_RUNNING = 2,
	STATE_STOPPED = 3,
	STATE_STOPPING = 4,
	STATE_RESTORING = 5,
	STATE_CHECKPOINTING = 6,
};

/* Parse error codes */
#define ERR_DUP		-1
#define ERR_INVAL	-2
#define ERR_UNK		-3
#define ERR_NOMEM	-4
#define ERR_OTHER	-5
#define ERR_INVAL_SKIP	-6
#define ERR_LONG_TRUNC  -7

/** CT handler.
 */
typedef struct {
	int vzfd;	/**< /dev/vzctl file descriptor. */
	int stdfd;
} vps_handler;

typedef enum {
	SKIP_NONE =		0,
	SKIP_SETUP =		(1<<0),
	SKIP_CONFIGURE =	(1<<1),
	SKIP_ACTION_SCRIPT =	(1<<2)
} skipFlags;

typedef int (* execFn)(void *data);

#endif


include/vzsyscalls.h
/*
 *  Copyright (C) 2000-2008, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
#ifndef _VZSYSCALLS_H_
#define _VZSYSCALLS_H_

#include <sys/syscall.h>

#ifdef __ia64__
#define __NR_fairsched_vcpus	1499
#define __NR_fairsched_chwt	1502
#define __NR_fairsched_rate	1504
#define __NR_setluid		1506
#define __NR_setublimit		1507
#define __NR_ioprio_set		1274
#elif __x86_64__
#define __NR_fairsched_vcpus	499
#define __NR_setluid		501
#define __NR_setublimit		502
#define __NR_fairsched_chwt	506
#define __NR_fairsched_rate	508
#define __NR_ioprio_set		251
#elif __powerpc__
#define __NR_fairsched_chwt	402
#define __NR_fairsched_rate	404
#define __NR_fairsched_vcpus	405
#define __NR_setluid		411
#define __NR_setublimit		412
#define __NR_ioprio_set		273
#elif defined(__i386__) || defined(__sparc__)
#define __NR_fairsched_chwt	502
#define __NR_fairsched_rate	504
#define __NR_fairsched_vcpus	505
#define __NR_setluid		511
#define __NR_setublimit		512
#ifdef __sparc__
#define __NR_ioprio_set		196
#else
#define __NR_ioprio_set		289
#endif
#else
#error "no syscall for this arch"
#endif

#endif


include/linux/vzcalluser.h
/*
 *  Copyright (C) 2000-2008, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef _LINUX_VZCALLUSER_H
#define _LINUX_VZCALLUSER_H

#include <linux/types.h>
#include <linux/ioctl.h>

#define KERN_VZ_PRIV_RANGE 51

#ifndef __ENVID_T_DEFINED__
typedef unsigned envid_t;
#define __ENVID_T_DEFINED__
#endif

#ifndef __KERNEL__
#define __user
#endif

/*
 * CT management ioctls
 */

struct vzctl_old_env_create {
	envid_t veid;
	unsigned flags;
#define VE_CREATE 	1	/* Create CT, VE_ENTER added automatically */
#define VE_EXCLUSIVE	2	/* Fail if exists */
#define VE_ENTER	4	/* Enter existing CT */
#define VE_TEST		8	/* Test if CT exists */
#define VE_LOCK		16	/* Do not allow entering created CT */
#define VE_SKIPLOCK	32	/* Allow entering "embryo" CT */
	__u32 addr;
};

struct vzctl_mark_env_to_down {
	envid_t veid;
};

struct vzctl_setdevperms {
	envid_t veid;
	unsigned type;
#define VE_USE_MAJOR	010	/* Test MAJOR supplied in rule */
#define VE_USE_MINOR	030	/* Test MINOR supplied in rule */
#define VE_USE_MASK	030	/* Testing mask, VE_USE_MAJOR|VE_USE_MINOR */
	unsigned dev;
	unsigned mask;
};

struct vzctl_ve_netdev {
	envid_t veid;
	int op;
#define VE_NETDEV_ADD  1
#define VE_NETDEV_DEL  2
	char __user *dev_name;
};

struct vzctl_ve_meminfo {
	envid_t veid;
	unsigned long val;
};

/* these masks represent modules */
#define VE_IP_IPTABLES_MOD		(1U<<0)
#define VE_IP_FILTER_MOD		(1U<<1)
#define VE_IP_MANGLE_MOD		(1U<<2)
#define VE_IP_MATCH_LIMIT_MOD		(1U<<3)
#define VE_IP_MATCH_MULTIPORT_MOD	(1U<<4)
#define VE_IP_MATCH_TOS_MOD		(1U<<5)
#define VE_IP_TARGET_TOS_MOD		(1U<<6)
#define VE_IP_TARGET_REJECT_MOD		(1U<<7)
#define VE_IP_TARGET_TCPMSS_MOD		(1U<<8)
#define VE_IP_MATCH_TCPMSS_MOD		(1U<<9)
#define VE_IP_MATCH_TTL_MOD		(1U<<10)
#define VE_IP_TARGET_LOG_MOD		(1U<<11)
#define VE_IP_MATCH_LENGTH_MOD		(1U<<12)
#define VE_IP_CONNTRACK_MOD		(1U<<14)
#define VE_IP_CONNTRACK_FTP_MOD		(1U<<15)
#define VE_IP_CONNTRACK_IRC_MOD		(1U<<16)
#define VE_IP_MATCH_CONNTRACK_MOD	(1U<<17)
#define VE_IP_MATCH_STATE_MOD		(1U<<18)
#define VE_IP_MATCH_HELPER_MOD		(1U<<19)
#define VE_IP_NAT_MOD			(1U<<20)
#define VE_IP_NAT_FTP_MOD		(1U<<21)
#define VE_IP_NAT_IRC_MOD		(1U<<22)
#define VE_IP_TARGET_REDIRECT_MOD	(1U<<23)
#define VE_IP_MATCH_OWNER_MOD		(1U<<24)
#define VE_IP_MATCH_MAC_MOD		(1U<<25)

/* these masks represent modules with their dependences */
#define VE_IP_IPTABLES		(VE_IP_IPTABLES_MOD)
#define VE_IP_FILTER		(VE_IP_FILTER_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_MANGLE		(VE_IP_MANGLE_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_LIMIT	(VE_IP_MATCH_LIMIT_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_MULTIPORT	(VE_IP_MATCH_MULTIPORT_MOD	\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_TOS		(VE_IP_MATCH_TOS_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_TARGET_TOS	(VE_IP_TARGET_TOS_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_TARGET_REJECT	(VE_IP_TARGET_REJECT_MOD	\
					| VE_IP_IPTABLES)
#define VE_IP_TARGET_TCPMSS	(VE_IP_TARGET_TCPMSS_MOD	\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_TCPMSS	(VE_IP_MATCH_TCPMSS_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_TTL		(VE_IP_MATCH_TTL_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_TARGET_LOG	(VE_IP_TARGET_LOG_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_LENGTH	(VE_IP_MATCH_LENGTH_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_CONNTRACK		(VE_IP_CONNTRACK_MOD		\
					| VE_IP_IPTABLES)
#define VE_IP_CONNTRACK_FTP	(VE_IP_CONNTRACK_FTP_MOD	\
					| VE_IP_CONNTRACK)
#define VE_IP_CONNTRACK_IRC	(VE_IP_CONNTRACK_IRC_MOD	\
					| VE_IP_CONNTRACK)
#define VE_IP_MATCH_CONNTRACK	(VE_IP_MATCH_CONNTRACK_MOD	\
					| VE_IP_CONNTRACK)
#define VE_IP_MATCH_STATE	(VE_IP_MATCH_STATE_MOD		\
					| VE_IP_CONNTRACK)
#define VE_IP_MATCH_HELPER	(VE_IP_MATCH_HELPER_MOD		\
					| VE_IP_CONNTRACK)
#define VE_IP_NAT		(VE_IP_NAT_MOD			\
					| VE_IP_CONNTRACK)
#define VE_IP_NAT_FTP		(VE_IP_NAT_FTP_MOD		\
					| VE_IP_NAT | VE_IP_CONNTRACK_FTP)
#define VE_IP_NAT_IRC		(VE_IP_NAT_IRC_MOD		\
					| VE_IP_NAT | VE_IP_CONNTRACK_IRC)
#define VE_IP_TARGET_REDIRECT	(VE_IP_TARGET_REDIRECT_MOD	\
					| VE_IP_NAT)
#define VE_IP_MATCH_MAC		(VE_IP_MATCH_MAC_MOD	\
					| VE_IP_IPTABLES)
#define VE_IP_MATCH_OWNER	(VE_IP_MATCH_OWNER_MOD	\
					| VE_IP_IPTABLES)

/* safe iptables mask to be used by default */
#define VE_IP_DEFAULT					\
	(VE_IP_IPTABLES |				\
	VE_IP_FILTER | VE_IP_MANGLE |			\
	VE_IP_MATCH_LIMIT | VE_IP_MATCH_MULTIPORT |	\
	VE_IP_MATCH_TOS | VE_IP_TARGET_REJECT | 	\
	VE_IP_TARGET_TCPMSS | VE_IP_MATCH_TCPMSS |	\
	VE_IP_MATCH_TTL | VE_IP_MATCH_LENGTH)

#define VE_IPT_CMP(x,y)		(((x) & (y)) == (y))

struct vzctl_env_create_cid {
	envid_t veid;
	unsigned flags;
	__u32 class_id;
};

struct vzctl_env_create {
	envid_t veid;
	unsigned flags;
	__u32 class_id;
};

struct env_create_param {
	__u64 iptables_mask;
};

#define VZCTL_ENV_CREATE_DATA_MINLEN	sizeof(struct env_create_param)

struct env_create_param2 {
	__u64 iptables_mask;
	__u64 feature_mask;
	__u32 total_vcpus;	/* 0 - don't care, same as in host */
};

struct env_create_param3 {
	__u64 iptables_mask;
	__u64 feature_mask;
	__u32 total_vcpus;
	__u32 pad;
	__u64 known_features;
};

#define VE_FEATURE_SYSFS	(1ULL << 0)
#define VE_FEATURE_NFS		(1ULL << 1)
#define VE_FEATURE_DEF_PERMS	(1ULL << 2)
#define VE_FEATURE_SIT		(1ULL << 3)
#define VE_FEATURE_IPIP		(1ULL << 4)

#define VE_FEATURES_OLD		(VE_FEATURE_SYSFS)
#define VE_FEATURES_DEF		(VE_FEATURE_SYSFS | \
				 VE_FEATURE_DEF_PERMS)

typedef struct env_create_param3 env_create_param_t;
#define VZCTL_ENV_CREATE_DATA_MAXLEN	sizeof(env_create_param_t)

struct vzctl_env_create_data {
	envid_t veid;
	unsigned flags;
	__u32 class_id;
	env_create_param_t __user *data;
	int datalen;
};

struct vz_load_avg {
	int val_int;
	int val_frac;
};

struct vz_cpu_stat {
	unsigned long user_jif;
	unsigned long nice_jif;
	unsigned long system_jif;
	unsigned long uptime_jif;
	__u64 idle_clk;
	__u64 strv_clk;
	__u64 uptime_clk;
	struct vz_load_avg avenrun[3];	/* loadavg data */
};

struct vzctl_cpustatctl {
	envid_t veid;
	struct vz_cpu_stat __user *cpustat;
};

#define VZCTLTYPE '.'
#define VZCTL_OLD_ENV_CREATE	_IOW(VZCTLTYPE, 0,			\
					struct vzctl_old_env_create)
#define VZCTL_MARK_ENV_TO_DOWN	_IOW(VZCTLTYPE, 1,			\
					struct vzctl_mark_env_to_down)
#define VZCTL_SETDEVPERMS	_IOW(VZCTLTYPE, 2,			\
					struct vzctl_setdevperms)
#define VZCTL_ENV_CREATE_CID	_IOW(VZCTLTYPE, 4,			\
					struct vzctl_env_create_cid)
#define VZCTL_ENV_CREATE	_IOW(VZCTLTYPE, 5,			\
					struct vzctl_env_create)
#define VZCTL_GET_CPU_STAT	_IOW(VZCTLTYPE, 6,			\
					struct vzctl_cpustatctl)
#define VZCTL_ENV_CREATE_DATA	_IOW(VZCTLTYPE, 10,			\
					struct vzctl_env_create_data)
#define VZCTL_VE_NETDEV		_IOW(VZCTLTYPE, 11,			\
					struct vzctl_ve_netdev)
#define VZCTL_VE_MEMINFO	_IOW(VZCTLTYPE, 13,                     \
					struct vzctl_ve_meminfo)

#ifdef __KERNEL__
#include <linux/compat.h>
#ifdef CONFIG_COMPAT
struct compat_vzctl_ve_netdev {
	envid_t veid;
	int op;
	compat_uptr_t dev_name;
};

struct compat_vzctl_ve_meminfo {
	envid_t veid;
	compat_ulong_t val;
};

struct compat_vzctl_env_create_data {
	envid_t veid;
	unsigned flags;
	__u32 class_id;
	compat_uptr_t data;
	int datalen;
};

#define VZCTL_COMPAT_ENV_CREATE_DATA _IOW(VZCTLTYPE, 10,		\
					struct compat_vzctl_env_create_data)
#define VZCTL_COMPAT_VE_NETDEV	_IOW(VZCTLTYPE, 11,			\
					struct compat_vzctl_ve_netdev)
#define VZCTL_COMPAT_VE_MEMINFO	_IOW(VZCTLTYPE, 13,                     \
					struct compat_vzctl_ve_meminfo)
#endif
#endif

#endif


pam_vz.c
/*
 * Copyright 2009 by Marco d'Itri <md@linux.it>.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <syslog.h>
#include <sys/types.h>
#include <errno.h>
#include <pwd.h>

#define PAM_SM_SESSION
#include <security/pam_modules.h>
#include <security/pam_ext.h>
#include <security/pam_modutil.h>

struct pam_vz_config {
    int min_id;
    int max_id;
    int use_gid;
    int ctid_delta;
    int closelog;
};

/* Prototypes */
extern int ve_enter(const id_t ctid);

static int doit(pam_handle_t *pamh, struct pam_vz_config *config);
static gid_t get_user_group_id(pam_handle_t *pamh, const int use_gid);
static int parse_option(pam_handle_t *pamh, struct pam_vz_config *config,
	const char *argv);
static int parse_options(pam_handle_t *pamh, struct pam_vz_config *config,
	const int argc, const char **argv);

/****************************************************************************/
/* Configuration defaults */
static struct pam_vz_config default_config = {
    .min_id = 10000,	/* totally arbitrary */
    .max_id = 65533,	/* (nobody - 1) */
    .use_gid = 0,	/* by default use the UID as CTID */
    .ctid_delta = 0,
    .closelog = 0,	/* close the syslog socket after entering the CT */
};

/* global flag used by log_d() */
static int debug = 0;

/* Convenience macros */
#define streq(a, b) (strcmp(a, b) == 0)
#define strneq(a, b, n) (strncmp(a, b, n) == 0)
#define strneq2(a, b) (strncmp(a, b, strlen(b)) == 0)

#define log_d(x, ...) do { \
    if (debug) \
	 pam_syslog(pamh, LOG_DEBUG,  x, ## __VA_ARGS__); \
} while(0);

#define log_n(x, ...) \
    do { pam_syslog(pamh, LOG_NOTICE, x, ## __VA_ARGS__); } while(0);

#define log_e(x, ...) \
    do { pam_syslog(pamh, LOG_ERR,    x, ## __VA_ARGS__); } while(0);

/****************************************************************************/
PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
			    const char **argv)
{
    struct pam_vz_config *config = &default_config;
    int rc;

    log_d("pam_sm_open_session called");

    rc = parse_options(pamh, config, argc, argv);
    if (rc != PAM_SUCCESS)
	return rc;

    if (!doit(pamh, config))
	return PAM_ABORT;

    return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_close_session(pam_handle_t *pamh, int flags, int argc,
			     const char **argv)
{
    log_d("pam_sm_close_session called");

    return PAM_IGNORE;
}

/****************************************************************************/
static int doit(pam_handle_t *pamh, struct pam_vz_config *config) {
    int rc;
    id_t ctid;

    /* Get the GID corresponding to the user opening the session.
     * -1 is used to report an error condition in the hope that nobody
     * would want to associate it to a CTID...
     */
    ctid = get_user_group_id(pamh, config->use_gid);
    if (ctid == -1)
	return 0;

    /* if the ID is not in the range then do nothing and return success */
    if (ctid < config->min_id || ctid > config->max_id) {
	log_d("%s %u is not in the CTIDs range, not entering a container",
		config->use_gid ? "gid" : "uid", ctid);
	return 1;
    }

    ctid += config->ctid_delta;

    log_n("trying to enter container %u", ctid);

    rc = ve_enter(ctid);

    /* Beware: in normal conditions the precedent call to syslog will cause
     * /dev/log in the HN to be opened and continue to be used even after
     * entering the container.
     * If the closelog option is specified then it will be closed and
     * after this point the process will try to log in the container (and
     * fail if it does not have a running syslog daemon).
     */
    if (rc == 1 && config->closelog)
	closelog();

    if (rc == 1) {
	log_n("entered container %u", ctid);
    } else
	log_e("failed entering container %u", ctid);

    return rc;
}

/****************************************************************************/
static gid_t get_user_group_id(pam_handle_t *pamh, const int use_gid)
{
    const char *user;
    struct passwd const *pw;
    int ret;

    ret = pam_get_item(pamh, PAM_USER, (const void **) &user);

    if (ret != PAM_SUCCESS) {
	log_e("pam_get_item(PAM_USER): %s", pam_strerror(pamh, ret));
	return -1;
    }

    errno = 0;
    pw = pam_modutil_getpwnam(pamh, user);
    if (!pw) {
	log_e("getpwnam(%s): %m", user);
	return -1;
    }

    return use_gid ? pw->pw_gid : pw->pw_uid;
}

/****************************************************************************/
static int parse_option(pam_handle_t *pamh, struct pam_vz_config *config,
	const char *argv)
{
    if        (streq(argv, "debug")) {
	debug = 1;
    } else if (strneq2(argv, "min_id=")) {
	errno = 0;
	config->min_id = strtol(argv + 8, NULL, 10);
	if (errno == ERANGE || errno == EINVAL || config->min_id == 0) {
	    log_e("unparseable value: %s", argv);
	    return PAM_SESSION_ERR;
	}
    } else if (strneq2(argv, "max_id=")) {
	errno = 0;
	config->max_id = strtol(argv + 7, NULL, 10);
	if (errno == ERANGE || errno == EINVAL || config->max_id == 0) {
	    log_e("unparseable value: %s", argv);
	    return PAM_SESSION_ERR;
	}
    } else if (strneq2(argv, "use_as_ctid=")) {
	if        (streq(argv + 12, "uid")) {
	    config->use_gid = 0;
	} else if (streq(argv + 12, "gid")) {
	    config->use_gid = 1;
	} else {
	    log_e("invalid argument: %s", argv);
	    return PAM_SESSION_ERR;
	}
    } else if (strneq2(argv, "ctid_delta=")) {
	errno = 0;
	config->ctid_delta = strtol(argv + 7, NULL, 10);
	if (errno == ERANGE || errno == EINVAL) {
	    log_e("unparseable value: %s", argv);
	    return PAM_SESSION_ERR;
	}
    } else if (streq(argv, "closelog")) {
	config->closelog = 1;
    } else {
	log_e("invalid config option: %s", argv);
	return PAM_SESSION_ERR;
    }
    return PAM_SUCCESS;
}

/* parse the module configuration directives */
static int parse_options(pam_handle_t *pamh, struct pam_vz_config *config,
	const int argc, const char **argv)
{
    int i;

    for (i = 0; i < argc; i++) {
	if (argv == NULL || argv[0] == '\0')
	    continue;

	int rc = parse_option(pamh, config, argv[i]);
	if (rc != PAM_SUCCESS)
	    return rc;
    }

    return PAM_SUCCESS;
}

/****************************************************************************/
#ifdef PAM_STATIC
/* used when the module is built static */
struct pam_module _pam_vz_modstruct = {
    "pam_vz",
    NULL,
    NULL,
    NULL,
    pam_sm_open_session,
    pam_sm_close_session,
    NULL,
};
#endif

